R/estimatr_lh_robust.R

```diff
@@ -0,0 +1,184 @@
+#'   z = rbinom(N, 1, prob = 0.4),
+#'   clusterID = sample(1:4, 40, replace = TRUE)
+#' )
+#'
+#' # Default variance estimator is HC2 robust standard errors
+#' lhro <- lh_robust(y ~ x + z, data = dat, linearHypothesis = "z + 2x = 0")
+#'
+#' # Also recovers other sorts of standard erorrs just as specified in \code{\link{lm_robust}}
+#' lh_robust(y ~ x + z, data = dat, linearHypothesis = "z + 2x = 0", se_type = "classical")
+#' lh_robust(y ~ x + z, data = dat, linearHypothesis = "z + 2x = 0", se_type =  "HC1")
+#'
+#' #  Can tidy() main output and subcomponents in to a data.frame
+#' lhro <- lh_robust(y ~ x + z, data = dat, linearHypothesis = "z + 2x = 0")
+#' tidy(lhro )
+#' tidy(lhro$lm_robust)
+#' tidy(lhro$lh)
+#'
+#' # Can use summary() to get more statistics on the main output and subcomponents.
+#' summary(lhro)
+#' summary(lhro$lm_robust)
+#' summary(lhro$lh)
+#'
+#' @export
+#'
+lh_robust <- function(formula,
+                      data,
+                      weights,
+                      subset,
+                      clusters,
+                      fixed_effects,
+                      se_type = NULL,
+                      ci = TRUE,
+                      alpha = .05,
+                      return_vcov = TRUE,
+                      try_cholesky = FALSE,
+                      linearHypothesis = NULL
+                      ) {
+
+  datargs <- enquos(
+    formula = formula,
+    weights = weights,
+    subset = subset,
+    cluster = clusters,
+    fixed_effects = fixed_effects
+  )
+  data <- enquo(data)
+  model_data <- clean_model_data(data = data, datargs)
+
+  fes <- is.character(model_data[["fixed_effects"]])
+  if (fes) {
+    yoriginal <- model_data[["outcome"]]
+    Xoriginal <- model_data[["design_matrix"]]
+    model_data <- demean_fes(model_data)
+    attr(model_data$fixed_effects, "fe_rank") <- sum(model_data[["fe_levels"]]) + 1
+  } else {
+    Xoriginal <- NULL
+    yoriginal <- NULL
+  }
+
+  return_list <-
+    lm_robust_fit(
+      y = model_data$outcome,
+      X = model_data$design_matrix,
+      yoriginal = yoriginal,
+      Xoriginal = Xoriginal,
+      weights = model_data$weights,
+      cluster = model_data$cluster,
+      fixed_effects = model_data$fixed_effects,
+      ci = ci,
+      se_type = se_type,
+      alpha = alpha,
+      return_vcov = return_vcov,
+      try_cholesky = try_cholesky,
+      has_int = attr(model_data$terms, "intercept"),
+      iv_stage = list(0)
+    )
+
+  return_list <- lm_return(
+    return_list,
+    model_data = model_data,
+    formula = formula
+  )
+
+  return_list[["call"]] <- match.call()
+
+
+  if(is.null(linearHypothesis)){
+    warning("No linear hypothesis test performed")
+    return( return_list)}
+
+  car_lht <- car::linearHypothesis(return_list, linearHypothesis,
+                               level = 1-alpha)
+
+  estimate  <- drop(attr(car_lht, "value"))
+  std.error <- sqrt(diag(attr(car_lht, "vcov")))
+  df <- return_list$df.residual
```

Comment:
Are these the right degrees of freedom for tests with, say, the CR2, satterthwaite-like degrees of freedom?

Reply:
Thanks for pointing at this. I certainly need to look a bit more carefully into it.

Though my sense would be that since this is a post-estimation analysis the degree of freedom should be the same as those used by the model.